From 559cfb6354da158d36e9331e419316aad3dbb3bc Mon Sep 17 00:00:00 2001
From: John Galt <johngaltfirstrun@gmail.com>
Date: Tue, 16 Dec 2025 09:46:08 -0500
Subject: [PATCH] private/domain: add new attr for relaxing a dir init
 neverallow

Will only be used by perspectived which is already a permissive domain
---
 prebuilts/api/202504/public/attributes | 3 +++
 private/domain.te                      | 3 ++-
 public/attributes                      | 3 +++
 3 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/prebuilts/api/202504/public/attributes b/prebuilts/api/202504/public/attributes
index ae7720ed9..46b6f2e0d 100644
--- a/prebuilts/api/202504/public/attributes
+++ b/prebuilts/api/202504/public/attributes
@@ -61,6 +61,9 @@ attribute vendor_file_type;
 attribute proc_type;
 expandattribute proc_type false;
 
+# Allowlisted domains that may violate the domain:dir restrictions on init
+attribute init_dir_mutation_exempt;
+
 # Types in /proc/net, excluding qtaguid types.
 # TODO(b/9496886) Lock down access to /proc/net.
 # This attribute is used to audit access to proc_net. it is temporary and will
diff --git a/private/domain.te b/private/domain.te
index bdc185c73..b45e39275 100644
--- a/private/domain.te
+++ b/private/domain.te
@@ -2363,6 +2363,7 @@ neverallow domain *:system x509_certificate_load;
 # generates SELinux denials prior to being stopped by other parts of
 # the kernel. Disallow such nonsensical allow rules
 neverallow domain proc:filesystem associate;
-neverallow domain domain:dir { write add_name remove_name };
+neverallow { domain -init_dir_mutation_exempt } domain:dir { write add_name remove_name };
+neverallow init_dir_mutation_exempt { domain -init }:dir {write add_name remove_name};
 neverallow domain domain:file create;
 
diff --git a/public/attributes b/public/attributes
index 274160f91..f610257fb 100644
--- a/public/attributes
+++ b/public/attributes
@@ -61,6 +61,9 @@ attribute vendor_file_type;
 attribute proc_type;
 expandattribute proc_type false;
 
+# Allowlisted domains that may violate the domain:dir restrictions on init
+attribute init_dir_mutation_exempt;
+
 # Types in /proc/net, excluding qtaguid types.
 # TODO(b/9496886) Lock down access to /proc/net.
 # This attribute is used to audit access to proc_net. it is temporary and will
-- 
2.51.2

